File: site/docs/tbdex/tests/issuer/kbc/knownBusinessCredentialIssuer.test.js
import { describe, test, expect, beforeAll } from 'vitest';
import { VerifiableCredential } from '@web5/credentials';
import { DidDht } from '@web5/dids';
import { Offering } from '@tbdex/http-client';

describe("Known Business Credential", () => {
    let pfiDid;
    let subjectDidUri;
    beforeAll(async () => {
        pfiDid = await DidDht.create({
            options: {
                publish: true,
                services: [{
                    id: 'pfi',
                    type: 'PFI',
                    serviceEndpoint: 'https://example.com/'
                }]
            },
        })
        const subjectDid = await DidDht.create()
        subjectDidUri = subjectDid.uri;
    })

    test("Issue Known Business Credential", async () => {
        // :snippet-start: issueKbcJs
        const kbc = await VerifiableCredential.create({
            issuer: pfiDid.uri, // Issuer's DID URI
            subject: subjectDidUri, // Wallet app's DID URI
            expirationDate: '2025-09-30T12:34:56Z', // Date the KBC should expire
            data: {
                //Custom attributes for the KBC 
                },
            credentialSchema: {
                type: "JsonSchema", // Format type of the schema used for the KBC
                id: "https://vc.schemas.host/kbc.schema.json" // URL to the schema used for the KBC
            }
        })
        const signedKbc = await kbc.sign({ did: pfiDid });
        // :snippet-end:
        expect(kbc).toBeDefined();
        expect.soft(kbc).toHaveProperty('issuer', pfiDid.uri);
        expect(typeof signedKbc).toBe('string');

    })

    test("Required claims in Known Business Credential", async () => {
        // :snippet-start: kbcPresentationDefinitionJs
        const pd = {
            id: "presentation-definition-kbc", // required unique id for presentation definition
            name: "KYB Verification",
            purpose: "Verifiying your business status.",
            format: {
                jwt_vc: {
                    alg: ["ES256K", "EdDSA"]
                }
            },
            input_descriptors: [
                {
                    id: "known-business-credential_1", // required unique id for the input descriptor
                    name: "Known Business Credential",
                    purpose: "Please present your Known Business Credential for verification.",
                    constraints: {
                        fields: [
                            {
                                path: ["$.vc.credentialSchema[0].id"],
                                filter: {
                                    type: "string",
                                    const: "https://vc.schemas.host/kbc.schema.json"
                                }
                            },
                            {
                                path: ["$.vc.issuer"],
                                filter: {
                                    type: "string",
                                    const: pfiDid.uri
                                }
                            }
                        ]
                    },
                }
            ]
        };
        // :snippet-end:

        try {
            // :snippet-start: kbcCreateOfferingJs
            const offering = Offering.create({
                metadata: {
                    from: pfiDid.uri,
                },
                data: {
                    description: "Selling BTC for USD",
                    payin: {
                        currencyCode: "USD",
                        max: "100.00",
                        methods: [
                          {
                            description: "Pay in via Debit Card, Apple Pay, or CashApp Pay",
                            kind: "PAYMENT_LINK",
                            name: "Debit Card, ApplePay, CashApp Pay"
                          }
                        ]
                    },
                    payout: {
                        currencyCode: 'BTC',
                        methods: [
                            {
                                kind: 'BTC_ADDRESS',
                                estimatedSettlementTime: 60,
                                fee: '0.25',
                            }
                        ]
                    },
                    payoutUnitsPerPayinUnit: '0.00003826',
                    // highlight-next-line
                    requiredClaims: pd
                }
            });

            await offering.sign(pfiDid);
            offering.validate();

            // :snippet-end:

        } catch (e) {
            expect.fail(e.message)
        }
    })
})

File: site-new/testsuites/testsuite-kotlin/src/test/kotlin/docs/tbdex/issuer/kbc/KnownBusinessCredentialIssuerTest.kt
package website.tbd.developer.site.docs.web5.build.verifiablecredentials

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import web5.sdk.credentials.VerifiableCredential
import web5.sdk.crypto.InMemoryKeyManager
import web5.sdk.dids.methods.dht.DidDht
import web5.sdk.dids.methods.dht.CreateDidDhtOptions
import web5.sdk.dids.didcore.Service
import java.time.Instant
import web5.sdk.credentials.CredentialSchema

import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.*
import tbdex.sdk.protocol.Parser
import tbdex.sdk.protocol.Validator
import tbdex.sdk.protocol.models.*
import web5.sdk.credentials.model.*
import java.util.*


/**
 * Tests backing the KBC Issuance Guide
 */
internal class KbcIssuanceTest {
    private val objectMapper = ObjectMapper()
   
    val serviceToAdd = Service.Builder()
        .id("pfi")
        .type("PFI")
        .serviceEndpoint(listOf("https://example.com/"))
        .build()

    val options = CreateDidDhtOptions(
        publish = true,
        services = listOf(serviceToAdd),
    )

    val pfiDid = DidDht.create(InMemoryKeyManager(), options)

    @Test
    fun `issue KBC`() {
        val subjectDid = DidDht.create(InMemoryKeyManager(),options)
        val subjectDidUri = subjectDid.uri
        data class AdditionalData(
            val exampleField: String = "exampleValue"
        )
        // :snippet-start: issueKbcKt
        val kbc = VerifiableCredential.create(
            issuer = pfiDid.uri, // Issuer's DID URI
            subject = subjectDidUri, // Wallet app's DID URI
            expirationDate = Date.from(Instant.parse("2025-09-30T12:34:56Z")), // Date the KBC should expire
            data = AdditionalData(), // Custom attributes for the KBC
            credentialSchema = CredentialSchema(
                type = "JsonSchema", // Format type of the schema used for the KBC
                id = "https://vc.schemas.host/kbc.schema.json" // URL to the schema used for the KBC
            )
        )
        val signedKbc = kbc.sign(pfiDid)
        // :snippet-end:
        assertNotNull(kbc)
        assertEquals(pfiDid.uri, kbc.issuer)
        assertEquals(subjectDidUri, kbc.subject)

        assertNotNull(signedKbc)
        assertTrue(signedKbc.matches(Regex("^[a-zA-Z0-9-_]+\\.[a-zA-Z0-9-_]+\\.[a-zA-Z0-9-_]+$")))
    }
    @Test
    fun `Required claims in Known Business Credential`() {
        // :snippet-start: kbcPresentationDefinitionKt
        val pd = PresentationDefinitionV2(
            id = "presentation-definition-kbc", // required unique id for presentation definition
            name = "KYB Verification",
            purpose = "Verifying your business status.",
            format = Format(
                jwtVc = JwtObject(
                    alg = listOf("ES256K", "EdDSA")
                )
            ),
            inputDescriptors = listOf(
                InputDescriptorV2(
                    id = "known-business-credential_1", // required unique id for the input descriptor
                    name = "Known Business Credential",
                    purpose = "Please present your Known Business Credential for verification.",
                    constraints = ConstraintsV2(
                        fields = listOf(
                            FieldV2(
                                path = listOf("$.credentialSchema.id"),
                                filterJson = objectMapper.readTree(
                                    """
                                    {
                                      "type": "string",
                                      "const": "https://vc.schemas.host/kbc.schema.json"
                                    }
                                    """.trimIndent()
                                ),
                            ),
                            FieldV2(
                                path = listOf("$.issuer"),
                                filterJson = objectMapper.readTree(
                                    """
                                    {
                                      "type": "string",
                                      "const":  "${pfiDid.uri}"
                                    }
                                    """.trimIndent()
                                )
                            )
                        )
                    )
                )
            )
        )
        // :snippet-end:

        try {
             // :snippet-start: kbcCreateOfferingKt
            val offering = Offering.create(
                from = pfiDid.uri,
                data = OfferingData(
                    description = "Selling BTC for USD",
                    payin = PayinDetails(
                        currencyCode = "USD",
                        max = "100.00",
                        methods = listOf(
                            PayinMethod(
                                description = "Pay in via Debit Card, Apple Pay, or CashApp Pay",
                                kind = "PAYMENT_LINK",
                                name = "Debit Card, ApplePay, CashApp Pay"
                            )
                        )
                    ),
                    payout = PayoutDetails(
                        currencyCode = "BTC",
                        methods = listOf(
                            PayoutMethod(
                                kind = "BTC_ADDRESS",
                                estimatedSettlementTime = 60,
                                fee = "0.25"
                            )
                        )
                    ),
                    payoutUnitsPerPayinUnit = "0.00003826",
                    // highlight-next-line
                    requiredClaims = pd
                )
            )

            offering.sign(pfiDid)
            Validator.validate(Parser.parseResourceToJsonNode(offering.toString()), "resource")
            // :snippet-end:
        } catch (e: Exception) {
            Assertions.fail(e.message)
        }
    }

}